src/test/java/jssc/junit/rules/VirtualPortRule.java

```diff
@@ -0,0 +1,187 @@
+    if (!osName.startsWith("linux")) {
+      return new Statement() {
+        @Override
+        public void evaluate() throws Throwable {
+          base.evaluate();
+        }
+      };
+    }
+
+    // is linux.
+
+    initUnix(description);
+
+    return new Statement() {
+
+      @Override
+      public void evaluate() throws Throwable {
+        try {
+          final BackgroundProcess socatProcess = openSocat();
+          final Future<?> socatFuture = executor.submit(socatProcess);
+          VirtualPortRule.this.processes.add(socatFuture);
+
+          // Java 8:
+          // System.timemillis…
+          // while (!socatProcess.getProcess().isAlive()) {
+          //  Thread.sleep(50);
+          // }
+
+          // java 6 way… pretend it is running.
+          Thread.sleep(500);
+
+          VirtualPortRule.this.isAvailable = true;
+
+          base.evaluate();
+        } finally {
+          // stop socat
+          for (final Future<?> process : VirtualPortRule.this.processes) {
+            process.cancel(false);
+          }
+
+          executor.shutdown();
+        }
+      }
+    };
+  }
+
+  private void initUnix(final Description description) {
+    final File testTargetDir = new File("./target").getAbsoluteFile();
+    final File virtualPortsDir = new File(testTargetDir, "virtual-ports");
+
+    virtualPortsDir.mkdirs();
+
+    final String filename1 = String
+        .format("virtualcom-%s-%s-%d", description.getClassName(), description.getMethodName(), this.random.nextInt(10000));
+    this.virtualCom1 = new File(virtualPortsDir, filename1);
+    final String fileName2 = String
+        .format("virtualcom-%s-%s-%d", description.getClassName(), description.getMethodName(), this.random.nextInt(10000));
+    this.virtualCom2 = new File(virtualPortsDir, fileName2);
+  }
+
+  private BackgroundProcess openSocat() {
+    return new BackgroundProcess() {
+
+      private Process process;
+
+      @Override
+      public Process getProcess() {
+        return this.process;
+      }
+
+      @Override
+      public void run() {
+
+        try {
+          final ProcessBuilder processBuilder = new ProcessBuilder(asList(
+              "/usr/bin/socat",
+              "pty,link=" + VirtualPortRule.this.virtualCom1.getAbsolutePath() + ",rawer,echo=0",
+              "pty,link=" + VirtualPortRule.this.virtualCom2.getAbsolutePath() + ",rawer,echo=0"
+          ));
+          processBuilder.redirectErrorStream(true);
+
+          this.process = processBuilder.start();
+          LOG.info("Process started! [{}], ports: [{}] // [{}].", this.process, VirtualPortRule.this.virtualCom1.getName(),
+              VirtualPortRule.this.virtualCom2.getName());
+          VirtualPortRule.this.outputStream = this.process.getOutputStream();
+          this.process.waitFor();
+        } catch (final IOException ioEx) {
+          throw new IllegalStateException("unable to start socat!", ioEx);
```

Comment:
Note, even without `socat` on my machine, I didn't get this error.

Reply:
Yes, the IOException will be thrown if the binary is available and cannot be started, iirc from the javadocs.

Reply:
No, it's not thrown.  Please re-read the comment.

Reply:
It won't compile without catching the IOException, because of `process.getOutputStream()` which is needed to read stdout of `socat`. There might be use cases where this is needed.

So at best, the error message is wrong or irritating. 
But if you insist, I can just remove the catch block and leave the code uncompileable.

Reply:
Some interest in reviving this has come up so I wanted to update my comment.

First, sorry for being rash in my previous comments.
Second, what I'd like to see is a usable error message if the binary is unavailable, which should be fairly sane request.

This comment is only provided for future contributors, it's not intended to be directed at the author specifically. ♥️

Reply:
Note taken.
You should get a more usable error message through the root cause. As the thrown exception is wrapping the original IOEx, it is not "swallowed". Instead, it is part of the stack trace.

I don't think there's much to do here, because the only way to get the reason why the call failed is to inspect the thrown IOEx.

So, here are two possible solutions:

* Throw IOEx directly, letting the library user deal with it. That might be a good idea for IOEx's anyway. Note that this is a breaking change though, as a new checked exception would be declared.
* Replace the message with `ioEx.getMessage()`. This duplicates the message (not recommended).
* Leave out the message (not recommended).

I would recommend the latter two options, because it's always good to have information about IN WHICH PART of the code the underlying exception happened.

I would worry about both the message and the API change, though, as it is only test code.

Happy coding!

Reply:
> Note, even without `socat` on my machine, I didn't get this error.

So there's something very bugged with `ProcessBuilder` in JDK6 mode.  I can't pinpoint the cause, but if `socat` is missing, it just deadlocks on `processBuilder.start()`.  For now, I'll just add a crude sanity check to make sure it exists before attempting to start it.

Reply:
Sanity check added.  Resolving.